stdlib/universe/kaufmansAMA.go

```diff
@@ -0,0 +1,301 @@
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+
+	return &KamaProcedureSpec{
+		N:      spec.N,
+		Column: spec.Column,
+	}, nil
+}
+
+func (s *KamaProcedureSpec) Kind() plan.ProcedureKind {
+	return kamaKind
+}
+
+func (s *KamaProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(KamaProcedureSpec)
+	*ns = *s
+	return ns
+}
+
+// TriggerSpec implements plan.TriggerAwareProcedureSpec
+func (s *KamaProcedureSpec) TriggerSpec() plan.TriggerSpec {
+	return plan.NarrowTransformationTriggerSpec{}
+}
+
+func createkamaTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*KamaProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewkamaTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type kamaTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	n      int64
+	column string
+
+	prevValue float64
+	currValue float64
+	prevKAMA  float64
+	sumUp     float64
+	sumDown   float64
+	diffNAgo  []float64 // keeps track of the last n diff values
+
+	count int64 // to see which data point we are currently on
+}
+
+func NewkamaTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *KamaProcedureSpec) *kamaTransformation {
+	return &kamaTransformation{
+		d:     d,
+		cache: cache,
+
+		n:      spec.N,
+		column: spec.Column,
+	}
+}
+
+func (t *kamaTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *kamaTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("KAMA found duplicate table with key: %v", tbl.Key())
+	}
+	if t.n <= 0 {
+		return fmt.Errorf("cannot take KaufmansAMA with a period of %v (must be greater than 0)", t.n)
+	}
+	cols := tbl.Cols()
+	doKAMA := make([]bool, len(cols))
+	for j, c := range cols {
+		found := false
+		if c.Label == t.column {
+			if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+				return fmt.Errorf("cannot take KAMA of column %s (type %s)", c.Label, c.Type.String())
+			}
+			found = true
+		}
+
+		if found {
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+			doKAMA[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	t.prevValue = 0
+	t.currValue = 0
+	t.prevKAMA = 0
+	t.sumUp = 0
+	t.sumDown = 0
+	t.diffNAgo = make([]float64, t.n)
+	t.count = 0
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 || len(cr.Cols()) == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			if !doKAMA[j] {
+				for i := int(t.n); i < cr.Len(); i++ {
```

Comment:
I'm not sure this will work quite right if there is more than one chunk in the table...  The first chunk will only have some rows copied to the output, but the latter chunks would need to have all their values copied.  You can find out if your on the first chunk by lookng at `t.count` I guess.